lib/puppetserver/ca/action/prune.rb

```diff
@@ -0,0 +1,123 @@
+
+Description:
+  Prune the list of revoked certificates of any duplication within it.
+
+Options:
+BANNER
+
+        def initialize(logger)
+          @logger = logger
+        end
+
+        def self.parser(parsed = {})
+          OptionParser.new do |opts|
+            opts.banner = BANNER # I will get a banner soon
+            opts.on('--help', 'Display this command-specific help output') do |help|
+              parsed['help'] = true
+            end
+            opts.on('--config CONF', 'Path to the puppet.conf file on disk') do |conf|
+              parsed['config'] = conf
+            end
+          end
+        end
+
+        # TODO:
+        # 1. Ensure that everything is offline (Done)
+        # 2. Get the path to the CRL (Done)
+        # 3. Prune each issuer revoked list (In progress)
+        # 4. Addin debug output
+        # 5. Repack the CRL and write it out?
+
+        def run(inputs)
+          config_path = inputs['config']
+
+
+          # Validate the config path.
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Validate puppet config setting.
+          settings_overrides = {}
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(settings_overrides)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Validate that we are offline
+          return 1 if HttpClient.check_server_online(puppet.settings, @logger)
+
+          # Getting the CRL(s)
+          loader = X509Loader.new(puppet.settings[:cacert], puppet.settings[:cakey], puppet.settings[:cacrl])
+
+          crl_list = loader.crls
+          prune_per_issuer(crl_list)  # Prune the list
+          return 0                  # Place holder return value for now
+        end
+
+        # Further note: The class Revoked override the == method.  We could make
+        # use of this by using array.uniq method to dedup ?
+        def prune_per_issuer(crl_list)
+          crl_list.each do |list|
+            existed_serial_number = Set.new()
+            revoked_list = list.revoked
+
+            revoked_list.delete_if do |revoked|
+              if existed_serial_number.add?(revoked.serial)
+                false
+              else
+                debug_output(list, revoked)
+                true
+              end
+            end
+            puts "Finished pruning CRL by #{list.issuer.to_s(OpenSSL::X509::Name::RFC2253)}\n"
+          end
+        end
+
+        # Output for debug
+        #  - Name of issuer
+        #  - Reason for revocation
+        #  - Time of revocation
+        def debug_output(curr_crl, curr_revoked_cert)
+          issuer_name = curr_crl.issuer.to_s(OpenSSL::X509::Name::RFC2253)
+          reason = curr_revoked_cert.extensions.join(', ') # Potentially ugly format.  Need more testing
+          time = curr_revoked_cert.time
```

Comment:
I don't think we need these. The cert will have already been revoked, so that info was probably logged elsewhere at the time of revocation. This is just removing duplicate entries from the CRL, so the issuer and cert subject are probably all you need for debug output?

Reply:
I tried to get the name of the cert but couldn't find a way to do so.  Apparently the `X509::Revoked` class does not have a getter for the cert name.  If we need the cert name, I will need to come up with a way to get that.

Reply:
hmm, based on the output you included, we could probably just use the alt names extension for that purpose.

Reply:
The `X509::Revoked.extensions` method doesn't return any name as far as I have tested.  It returns an array and the only thing in that array was the reason of revocation.  Maybe I've misunderstood what you mean by alt names extension?

Reply:
oh, I was looking at the alt names from
```
Revoked Certificates:
    tu             (SHA256)  18:52:07:6E:6F:E0:42:40:20:75:B8:80:0F:5D:26:94:0A:22:DC:2A:DD:67:81:90:BA:B3:13:8D:AD:B3:B5:95    alt names: ["DNS:tu"]
    batman         (SHA256)  8B:C6:E1:62:8B:E4:90:69:8D:0D:35:A4:36:CC:D8:EE:4B:74:89:06:80:49:21:F2:AD:AA:F9:63:DF:0D:20:42    alt names: ["DNS:batman"]
    superman       (SHA256)  52:3A:D2:06:26:AA:0C:EE:BA:B5:AB:5D:21:B5:12:4C:91:98:98:9A:23:B2:ED:EE:1C:B2:64:3D:67:39:27:AA    alt names: ["DNS:superman"]
```
but maybe that uses a different object